pkg/scheduler/framework/plugins/podtopologyspread/filtering.go

```diff
@@ -251,29 +256,27 @@ func (pl *PodTopologySpread) calPreFilterState(ctx context.Context, pod *v1.Pod)
 		TpPairToMatchNum:     make(map[topologyPair]int, sizeHeuristic(len(allNodes), constraints)),
 	}
 
-	requiredSchedulingTerm := nodeaffinity.GetRequiredNodeAffinity(pod)
 	tpCountsByNode := make([]map[topologyPair]int, len(allNodes))
+	requiredNodeAffinity := nodeaffinity.GetRequiredNodeAffinity(pod)
 	processNode := func(i int) {
 		nodeInfo := allNodes[i]
 		node := nodeInfo.Node()
 		if node == nil {
 			klog.ErrorS(nil, "Node not found")
 			return
 		}
-		// In accordance to design, if NodeAffinity or NodeSelector is defined,
-		// spreading is applied to nodes that pass those filters.
-		// Ignore parsing errors for backwards compatibility.
-		match, _ := requiredSchedulingTerm.Match(node)
-		if !match {
-			return
-		}
+
 		// Ensure current node's labels contains all topologyKeys in 'Constraints'.
 		if !nodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
 			return
 		}
 
 		tpCounts := make(map[topologyPair]int, len(constraints))
 		for _, c := range constraints {
+			if !c.matchHonorNodeInclusionPolicies(pod, node, requiredNodeAffinity) {
```

Comment:
this is being called inside a loop, where before, it was called outside... is that going to significantly change performance when the feature is enabled? what about when the feature is disabled?

Reply:
Yes, before we verify the nodeAffinity at `node` level, but now NodeInclusionPolicy performs at `constrain` level.
Considering `Constraints`'s amount in real cases will not be quite large, I think it's controllable. Here's the benchmark test results.
1. Master Branch: `Verify nodeAffinity and Skip taints verification`
2. Current Branch: 
   - Feature gate is disabled. `Verify nodeAffinity and Skip taints verification` with new implement. `No significantly performance downgrade` .
   - Feature gate enabled. We didn't run benchmark for this scenario. But the difference locates at the taints verification. Since it looks like there's no parsing involved in taints verification and the taints are limited, `I also think there's no significantly performance downgrade`. (I'll add some tests later to this benchmark to confirm this.)
   

Mater branch:
```
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkFilter/1000nodes/single-constraint-zone-8                  1458            686904 ns/op
BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8          808203              1693 ns/op
BenchmarkFilter/1000nodes/single-constraint-node-8                  1573            779965 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-8           12524             95071 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-8                938           1168812 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8                12606             95899 ns/op
PASS
ok      k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread     10.881s
```

Current Branch:
```
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkFilter/1000nodes/single-constraint-zone-8                  1682            737574 ns/op
BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8          790586              1589 ns/op
BenchmarkFilter/1000nodes/single-constraint-node-8                  1612            777061 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-8           12580             94916 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-8                942           1200677 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8                12678             95820 ns/op
PASS
ok      k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread     12.480s
``` 

We'll also add more benchmark tests when graduating to beta level.